src/cargo/core/resolver/mod.rs

```diff
@@ -1098,3 +1099,28 @@ fn check_cycles(resolve: &Resolve, activations: &Activations) -> CargoResult<()>
         Ok(())
     }
 }
+
+fn get_duplicate_pkgs(resolve: &Resolve) -> Vec<&'static str> {
+    let mut unique_pkg_ids = HashSet::new();
+    let mut result = HashSet::new();
+    for pkg_id in resolve.iter() {
+        let mut encodable_pkd_id = encode::encodable_package_id(pkg_id);
+        if !unique_pkg_ids.insert(encodable_pkd_id) {
+            result.insert(pkg_id.name().as_str());
+        }
+    }
+    result.into_iter().collect()
+}
+
+fn check_duplicate_pkgs(resolve: &Resolve) -> CargoResult<()> {
+    let names = get_duplicate_pkgs(resolve);
+    if names.is_empty() {
+        Ok(())
+    } else {
+        bail!(
+            "dependencies contain duplicate package(s) in the \
```

Comment:
I tried to come up with a short but precise message. Not sure this is good enough. Could be confusing for the user.

Reply:
Yeah, I think we could tweak this a little bit to print actual `PackageId`s, which include path and should hopefully make debugging this easier. 

Let's return a pair of `PackgeId` from `get_duplicate_pkgs`. Then the error message cold say something like:

```
bail!(
    "package collision in the lockfile: packages {} and {} are different,  but only one can be written to lockfile unambigiously",
    pkg_id_a,
    pkg_id_b,
)
```

Reply:
>I tried to come up with a short but precise message.

In case of rate and obscure errors it's better to write longer-than-usual messages :)

Reply:
Great! This makes much more sense.